### PR TITLE
Chroot into a new dir before starting the system

### PR DESCRIPTION
**Pull request by Itxaka**

### CODE REVIEW

File: pkg/mount/dag_steps.go [hunk 1]
```diff
@@ -808,3 +788,96 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+				internalUtils.Log.Debug().Str("out", output).Msg("systemd-pcrphase leave-initrd")
+			}
+			// Print dag before exit, otherwise its never printed as we never exit the program
+			internalUtils.Log.Info().Msg(s.WriteDAG(g))
+
+			// Mount a tmpfs under sysroot
+			err = syscall.Mount("tmpfs", s.path("sysroot"), "tmpfs", syscall.MS_NOSUID|syscall.MS_NODEV|syscall.MS_NOEXEC, "")
+
+			// Move all the dirs in root FS that are not a mountpoint to the new root via Bind mount
+			err = filepath.WalkDir(s.Rootdir, func(path string, d fs.DirEntry, err error) error {
```
[the remark this answers is not in the document]
Comment (mudler) [reply pkg/mount/dag_steps.go:812]: You can also just use ReadDir if top level only is needed

[the remark this answers is not in the document]
Comment (mudler) [reply pkg/mount/dag_steps.go:812]: https://pkg.go.dev/os#example-ReadDir

File: pkg/mount/dag_steps.go [hunk 2]
```diff
@@ -808,3 +788,96 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+				}
+				parentPath := filepath.Dir(s.path(path))
+				parentInfo, err := os.Stat(parentPath)
+				if err != nil {
+					return err
+				}
+				// If the directory has the same device as its parent, it's not a mount point.
+				if fileInfo.Sys().(*syscall.Stat_t).Dev == parentInfo.Sys().(*syscall.Stat_t).Dev {
+					internalUtils.Log.Info().Msg(fmt.Sprintf("%s is a simple directory", path))
+					err = os.MkdirAll(filepath.Join(s.path("sysroot"), path), 0755)
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:831]: image

Comment (jimmykarily) [reply pkg/mount/dag_steps.go:831]: first problem, it should never go recursively (only stick to `/dev`). Second, it won't be able to create directories anyway (?).

File: pkg/mount/dag_steps.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Itxaka) [reply pkg/mount/dag_steps.go]: redone and check for root files to be directly copied to new rootdir

**Review by jimmykarily — APPROVED** [review]

**Comment by jimmykarily** [comment]
I removed the `config.toml.tmpl` and restarted k3s. Then deleted Pods and saw them coming up just fine. Also installed cilium ([following the docs](https://docs.cilium.io/en/stable/gettingstarted/k8s-install-default/)) and it came up fine (well almost, the Pods start but we would probably have to remove flannel to make it work fully. But the `/proc` error is gone).

File: internal/utils/log.go [hunk 3]
```diff
@@ -32,6 +32,17 @@ func SetLogger() {
 		w.TimeFormat = time.RFC3339
 	}))
 
+	// Create a logger that logs to ttyS0
+	// Open the serial port for writing
+	serial, err := os.OpenFile("/dev/ttyS0", os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0666)
```
Comment (Itxaka) [inline internal/utils/log.go:37]: This should be removed as it doesn't work

**Comment by Itxaka** [comment]
Also most of the logs should be either dropped or moved to febug

File: pkg/mount/dag_steps.go [hunk 4]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+				internalUtils.Log.Err(err).Msg("mounting tmpfs on sysroot")
+			}
+
+			// Move all the dirs in root FS that are not a mountpoint to the new root via Bind mount
+			rootDirs, err := os.ReadDir(s.Rootdir)
+			if err != nil {
+				internalUtils.Log.Err(err).Msg("reading rootdir content")
+			}
+			mountPoints := []string{}
+			internalUtils.Log.Info().Str("s", litter.Sdump(rootDirs)).Msg("Moving root dirs to sysroot")
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:811]:
```suggestion
			internalUtils.Log.Debug().Str("s", litter.Sdump(rootDirs)).Msg("Moving root dirs to sysroot")
```

File: pkg/mount/dag_steps.go [hunk 5]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+						return err
+					}
+					parentPath := filepath.Dir(s.path(path))
+					parentInfo, err := os.Stat(parentPath)
+					if err != nil {
+						return err
+					}
+					// If the directory has the same device as its parent, it's not a mount point.
+					if fileInfo.Sys().(*syscall.Stat_t).Dev == parentInfo.Sys().(*syscall.Stat_t).Dev {
+						internalUtils.Log.Info().Msg(fmt.Sprintf("%s is a simple directory", path))
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:829]:
```suggestion
						internalUtils.Log.Debug().Msg(fmt.Sprintf("%s is a simple directory", path))
```

File: pkg/mount/dag_steps.go [hunk 6]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+							return err
+						}
+						// Bind mount it
+						err = syscall.Mount(s.path(path), filepath.Join(s.path(cnst.UkiSysrootDir), path), "", syscall.MS_BIND, "")
+						if err != nil {
+							internalUtils.Log.Err(err).Str("what", s.path(path)).
+								Str("where", filepath.Join(s.path(cnst.UkiSysrootDir), path)).Msg("bind mount")
+							return err
+						}
+						internalUtils.Log.Info().Msg(fmt.Sprintf("Bind mounted %s to %s", s.path(path), filepath.Join(s.path(cnst.UkiSysrootDir), path)))
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:842]:
```suggestion
						internalUtils.Log.Debug().Msg(fmt.Sprintf("Bind mounted %s to %s", s.path(path), filepath.Join(s.path(cnst.UkiSysrootDir), path)))
```

File: pkg/mount/dag_steps.go [hunk 7]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+
+						continue
+					}
+
+					internalUtils.Log.Info().Msg(fmt.Sprintf("%s is a mount point, skipping", s.path(path)))
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:847]:
```suggestion
					internalUtils.Log.Debug().Msg(fmt.Sprintf("%s is a mount point, skipping", s.path(path)))
```

File: pkg/mount/dag_steps.go [hunk 8]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+					target, err := os.Readlink(file.Name())
+					if err != nil {
+						return fmt.Errorf("failed to read symlink: %w", err)
+					}
+					symlinkPath := s.path(filepath.Join(cnst.UkiSysrootDir, file.Name()))
+					err = os.Symlink(target, symlinkPath)
+					if err != nil {
+						return fmt.Errorf("failed to create symlink: %w", err)
+					}
+					internalUtils.Log.Info().Str("from", target).Str("to", symlinkPath).Msg("Symlinked file")
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:867]:
```suggestion
					internalUtils.Log.Debug().Str("from", target).Str("to", symlinkPath).Msg("Symlinked file")
```

File: pkg/mount/dag_steps.go [hunk 9]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+				} else {
+					// If its a file in the root dir just copy it over
+					content, _ := os.ReadFile(s.path(file.Name()))
+					newFilePath := s.path(filepath.Join(cnst.UkiSysrootDir, file.Name()))
+					_ = os.WriteFile(newFilePath, content, info.Mode())
+					internalUtils.Log.Info().Msg(fmt.Sprintf("Copied %s to %s", s.path(file.Name()), newFilePath))
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:873]:
```suggestion
					internalUtils.Log.Debug().Msg(fmt.Sprintf("Copied %s to %s", s.path(file.Name()), newFilePath))
```

File: pkg/mount/dag_steps.go [hunk 10]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+				err := os.MkdirAll(newDir, 0755)
+				if err != nil {
+					internalUtils.Log.Err(err).Str("what", newDir).Msg("mkdir")
+				}
+				err = syscall.Mount(filepath.Join(s.path(), d), newDir, "", syscall.MS_MOVE, "")
+				if err != nil {
+					internalUtils.Log.Err(err).Str("what", filepath.Join(s.Rootdir, d)).Str("where", newDir).Msg("move mount")
+					continue
+				}
+				internalUtils.Log.Info().Str("from", filepath.Join(s.path(), d)).Str("to", newDir).Msg("Mount moved")
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:888]:
```suggestion
				internalUtils.Log.Debug().Str("from", filepath.Join(s.path(), d)).Str("to", newDir).Msg("Mount moved")
```

File: pkg/mount/dag_steps.go [hunk 11]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+			if err = syscall.Mount("", s.path(cnst.UkiSysrootDir), "", syscall.MS_REMOUNT|syscall.MS_RDONLY, ""); err != nil {
+				internalUtils.Log.Err(err).Msg("re-mounting sysroot as read only")
+			}
+
+			// Now chdir+chroot into the new dir
+			if err := unix.Chdir(s.path(cnst.UkiSysrootDir)); err != nil {
+				internalUtils.Log.Err(err).Msg("chdir")
+				return fmt.Errorf("failed change directory to new_root %v", err)
+			}
+			internalUtils.Log.Info().Msg("Chdir to sysroot done")
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:901]:
```suggestion
			internalUtils.Log.Debug().Msg("Chdir to sysroot done")
```

File: pkg/mount/dag_steps.go [hunk 12]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+
+			err = unix.Chroot(".")
+			if err != nil {
+				internalUtils.Log.Err(err).Msg("chroot")
+				return fmt.Errorf("failed to chroot %v", err)
+			}
+			internalUtils.Log.Info().Msg("Chroot to sysroot done")
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:908]:
```suggestion
			internalUtils.Log.Debug().Msg("Chroot to sysroot done")
```

File: pkg/mount/dag_steps.go [hunk 13]
```diff
@@ -808,3 +779,142 @@ func (s *State) MountLiveCd(g *herd.Graph, opts ...herd.OpOption) error {
...
+
+			internalUtils.Log.Info().Msg("Executing init callback!")
```
Comment (jimmykarily) [inline pkg/mount/dag_steps.go:910]:
```suggestion
			internalUtils.Log.Debug().Msg("Executing init callback!")
```

File: internal/utils/log.go (hunk 3 above)
Comment (jimmykarily) [reply internal/utils/log.go:37]: does it not? same thing for kmesg below?

**Comment by jimmykarily** [comment]
> Also most of the logs should be either dropped or moved to febug

done (moved to debug)

File: internal/utils/log.go (hunk 3 above)
Comment (Itxaka) [reply internal/utils/log.go:37]: I could only test it a couple of times but I think at this point the  /dev dir is not there and if it is it's not working as expected so I rather leave it to implement it properly later.

Plus we still got to move to the new Kairos logger :)

Comment (jimmykarily) [reply internal/utils/log.go]: done

**Review by jimmykarily — APPROVED** [review]

**Comment by jimmykarily** [comment]
I'm merging so that it gets tested more thoroughly on kairos repo

**Pull request merged**